File: iso27001-fastapi/app/core/brute_force.py
"""
A.9: Brute-force login protection.

Tracks failed authentication attempts per account identifier (email).
Uses Redis when available (cross-process, survives restarts); falls back to
an in-process dict for dev/test environments without Redis.

Policy:
  - MAX_ATTEMPTS  : 5 consecutive failures trigger a lockout
  - LOCKOUT_TTL   : 15 minutes (900 seconds)
  - Window resets on a successful login (clear())

Install Redis for production accuracy:
  REDIS_URL=redis://127.0.0.1:6379
"""

import time
from typing import Any
from fastapi import HTTPException, status

_MAX_ATTEMPTS: int = 5
_LOCKOUT_TTL: int = 900  # seconds (15 minutes)
_KEY_PREFIX = "brute_force:"

# ── in-process fallback storage ──────────────────────────────────────────────
# {email: {"count": int, "locked_until": float}}
_local: dict[str, dict[str, float | int]] = {}


def _redis_client() -> Any:
    """Return a redis.Redis client or None if unavailable."""
    try:
        import redis as _redis
        from app.config.settings import settings

        client = _redis.Redis.from_url(
            settings.REDIS_URL, encoding="utf-8", decode_responses=True, socket_connect_timeout=0.5
        )
        client.ping()
        return client
    except Exception:
        return None


class BruteForceGuard:
    """
    Thread-safe (via Redis atomics) brute-force guard.
    Degrades gracefully to in-process counters when Redis is unavailable.
    """

    def check(self, identifier: str) -> None:
        """Raise HTTP 429 if the account is currently locked out."""
        r = _redis_client()
        key_locked = f"{_KEY_PREFIX}{identifier}:locked_until"

        if r is not None:
            locked_until = r.get(key_locked)
            if locked_until and float(locked_until) > time.time():
                raise HTTPException(
                    status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                    detail={
                        "code": "ACCOUNT_LOCKED",
                        "message": "Too many failed attempts. Account temporarily locked.",
                    },
                )
        else:
            entry = _local.get(identifier)
            if entry and float(entry.get("locked_until", 0)) > time.time():
                raise HTTPException(
                    status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                    detail={
                        "code": "ACCOUNT_LOCKED",
                        "message": "Too many failed attempts. Account temporarily locked.",
                    },
                )

    def record_failure(self, identifier: str) -> None:
        """Increment failure counter; lock the account on threshold breach."""
        r = _redis_client()
        key_count = f"{_KEY_PREFIX}{identifier}:count"
        key_locked = f"{_KEY_PREFIX}{identifier}:locked_until"

        if r is not None:
            count = r.incr(key_count)
            r.expire(key_count, _LOCKOUT_TTL)
            if int(count) >= _MAX_ATTEMPTS:
                locked_until = time.time() + _LOCKOUT_TTL
                r.set(key_locked, locked_until, ex=_LOCKOUT_TTL)
                r.delete(key_count)
        else:
            entry = _local.setdefault(identifier, {"count": 0, "locked_until": 0.0})
            entry["count"] = int(entry["count"]) + 1
            if int(entry["count"]) >= _MAX_ATTEMPTS:
                entry["locked_until"] = time.time() + _LOCKOUT_TTL
                entry["count"] = 0

    def clear(self, identifier: str) -> None:
        """Clear failure counters after a successful login."""
        r = _redis_client()
        if r is not None:
            r.delete(
                f"{_KEY_PREFIX}{identifier}:count",
                f"{_KEY_PREFIX}{identifier}:locked_until",
            )
        else:
            _local.pop(identifier, None)


# Module-level singleton
brute_force_guard = BruteForceGuard()

File: iso27001-gin/internal/core/middleware/brute_force.go
package middleware

// A.9: Brute-force login protection.
//
// Tracks failed authentication attempts per account identifier (email).
// Uses Redis when available (cross-process); falls back to in-process map.
//
// Policy:
//   - MaxAttempts : 5 consecutive failures trigger a lockout
//   - LockoutTTL  : 15 minutes (900 seconds)
//   - Cleared     : on successful login via Clear()

import (
	"context"
	"fmt"
	"net/http"
	"strconv"
	"sync"
	"time"

	"github.com/gin-gonic/gin"
	goredis "github.com/redis/go-redis/v9"
)

const (
	bruteMaxAttempts = 5
	bruteLockoutTTL  = 15 * time.Minute
	bruteKeyPrefix   = "brute_force:"
)

type localEntry struct {
	count       int
	lockedUntil time.Time
}

// BruteForceGuard checks and records failed auth attempts.
type BruteForceGuard struct {
	rdb *goredis.Client
	mu  sync.Mutex
	mem map[string]*localEntry
}

// NewBruteForceGuard creates a guard; pass nil rdb to use in-process only.
func NewBruteForceGuard(rdb *goredis.Client) *BruteForceGuard {
	return &BruteForceGuard{rdb: rdb, mem: map[string]*localEntry{}}
}

// Check aborts with 429 if the identifier is currently locked out.
func (g *BruteForceGuard) Check(c *gin.Context, identifier string) bool {
	if g.isLocked(identifier) {
		c.AbortWithStatusJSON(http.StatusTooManyRequests, gin.H{
			"error": gin.H{
				"code":    "ACCOUNT_LOCKED",
				"message": "Too many failed attempts. Account temporarily locked.",
			},
		})
		return false
	}
	return true
}

// RecordFailure increments the failure counter, locking on threshold breach.
func (g *BruteForceGuard) RecordFailure(identifier string) {
	if g.rdb != nil {
		ctx, cancel := context.WithTimeout(context.Background(), 300*time.Millisecond)
		defer cancel()
		keyCount := fmt.Sprintf("%s%s:count", bruteKeyPrefix, identifier)
		keyLock := fmt.Sprintf("%s%s:locked_until", bruteKeyPrefix, identifier)
		count, _ := g.rdb.Incr(ctx, keyCount).Result()
		g.rdb.Expire(ctx, keyCount, bruteLockoutTTL)
		if count >= bruteMaxAttempts {
			g.rdb.Set(ctx, keyLock, time.Now().Add(bruteLockoutTTL).Unix(), bruteLockoutTTL)
			g.rdb.Del(ctx, keyCount)
		}
		return
	}
	g.mu.Lock()
	defer g.mu.Unlock()
	e := g.memEntry(identifier)
	e.count++
	if e.count >= bruteMaxAttempts {
		e.lockedUntil = time.Now().Add(bruteLockoutTTL)
		e.count = 0
	}
}

// Clear resets the failure state after a successful login.
func (g *BruteForceGuard) Clear(identifier string) {
	if g.rdb != nil {
		ctx, cancel := context.WithTimeout(context.Background(), 300*time.Millisecond)
		defer cancel()
		g.rdb.Del(ctx,
			fmt.Sprintf("%s%s:count", bruteKeyPrefix, identifier),
			fmt.Sprintf("%s%s:locked_until", bruteKeyPrefix, identifier),
		)
		return
	}
	g.mu.Lock()
	defer g.mu.Unlock()
	delete(g.mem, identifier)
}

func (g *BruteForceGuard) isLocked(identifier string) bool {
	if g.rdb != nil {
		ctx, cancel := context.WithTimeout(context.Background(), 300*time.Millisecond)
		defer cancel()
		val, err := g.rdb.Get(ctx, fmt.Sprintf("%s%s:locked_until", bruteKeyPrefix, identifier)).Result()
		if err != nil {
			return false
		}
		ts, err := strconv.ParseInt(val, 10, 64)
		if err != nil {
			return false
		}
		return time.Now().Unix() < ts
	}
	g.mu.Lock()
	defer g.mu.Unlock()
	e := g.mem[identifier]
	return e != nil && time.Now().Before(e.lockedUntil)
}

func (g *BruteForceGuard) memEntry(identifier string) *localEntry {
	if g.mem[identifier] == nil {
		g.mem[identifier] = &localEntry{}
	}
	return g.mem[identifier]
}
